### ISSUE TITLE
Enhance the python feature server with new `list` endpoints.

### ISSUE THREAD

**Issue by tchughesiv** [feast-dev/feast#4329]
**Is your feature request related to a problem? Please describe.**
With the python feature server, there should be a way to list various objects by type and filter them by tags.

**Describe the solution you'd like**
Feature server endpoints should be added for listing each of the following object types -
- DataSource
- Entity
- FeatureService
- FeatureView
- OnDemandFeatureView
- StreamFeatureView

The user should have the option of filtering by tags and/or bypassing cache where supported.

**Describe alternatives you've considered**
I considered adding tag filtering to the existing `/get-online-features` endpoint, but this would be insufficient.

**Additional context**
The feast UI & CLI already allow for filtering by tags.

**Comment by tokoko** [comment]
Isn't that the job of a registry server?

**Comment by tchughesiv** [comment]
> Isn't that the job of a registry server?

@tokoko yes, and it will still be the job of the registry server. we'd just add new endpoints in `feature_server.py` that leverage the existing registry methods.

**Comment by tokoko** [comment]
I'm not sure I'm a fan, tbh. We'll end up having virtually identical endpoints in [registry_server.py](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/registry_server.py) and [feature_server.py](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_server.py), the only difference being that one is a `grpc` server and the other's an `http` one. We'll also have to secure it twice to validate permissions (see this ongoing [PR](https://github.com/RHEcosystemAppEng/feast/pull/25) for context). The surface area / complexity that non-python feature servers will need to be implement would also increase...

Do you have some specific use case in mind?

**Comment by dmartinol** [comment]
> I'm not sure I'm a fan, tbh. We'll end up having virtually identical endpoints in [registry_server.py](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/registry_server.py) and [feature_server.py](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/feature_server.py) ...

Agree, the servers should implement the equivalent Feast interfaces like `OnlineStore`, `OfflineStore` and `Registry`, so that they can be transparently accessed by remote python clients w/o using any 3rd party library (`requests`, `pyarrow.flight`, `grpc` and so on). Just run `fs.registry.list_feature_views()`. Adding more endpoints would break this pattern.

**Comment by tchughesiv** [comment]
@tokoko @dmartinol the use-case for this enhancement is to allow for a potential future feast integration with the [odh-dashboard](https://github.com/opendatahub-io/odh-dashboard). based on some initial discussions i had, it appeared an http server was required. 

i've done some more digging this morning though and there appears to have been some more recent developments to the project around gRPC support. i'll continue my research and report back.

**Comment by dmartinol** [comment]
@tokoko just out of curiosity, what was the reason to develop the Registry Server using grpc instead of REST?

**Comment by dmartinol** [comment]
@tchughesiv is there a list of APIs requested to populate the dashboard and manage the exposed functions?
Maybe a dedicated Backend For Frontend service is more suitable instead of (trying to) use the endpoints of the various Feast servers which are instead targeted to serve the ML workflows. 
Of course, this BFF can use the Feast server endpoints, but with interfaces tailored to frontend needs (and yeah, it can also 'talk' gRPC and Arrow Flight-ish), and can be installed by the operator as it needs to refer to the other servers exposing the Feast features.

Oops, I was almost implementing it 🤫  EOM.

**Comment by tchughesiv** [comment]
Just to provide an update on the use of gRPC-web w/ the feast registry's gRPC API. My tests were successful, so updates to the feature server are not necessary.
https://github.com/grpc/grpc-web/compare/master...tchughesiv:grpc-web:dev
